pkg/cmd/checkendpoints/controller/controller.go

```diff
@@ -0,0 +1,255 @@
+package controller
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"time"
+
+	operatorcontrolplanev1alpha1 "github.com/openshift/api/operatorcontrolplane/v1alpha1"
+	operatorcontrolplaneclientv1alpha1 "github.com/openshift/client-go/operatorcontrolplane/clientset/versioned/typed/operatorcontrolplane/v1alpha1"
+	alpha1 "github.com/openshift/client-go/operatorcontrolplane/informers/externalversions/operatorcontrolplane/v1alpha1"
+	"github.com/openshift/client-go/operatorcontrolplane/listers/operatorcontrolplane/v1alpha1"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/klog"
+
+	"github.com/openshift/cluster-kube-apiserver-operator/pkg/cmd/checkendpoints/operatorcontrolplane/podnetworkconnectivitycheck/v1alpha1helpers"
+	"github.com/openshift/cluster-kube-apiserver-operator/pkg/cmd/checkendpoints/trace"
+)
+
+// PodNetworkConnectivityCheckController continuously performs network connectivity
+// checks and reports the results.
+type PodNetworkConnectivityCheckController interface {
+	factory.Controller
+}
+
+// controller implements a PodNetworkConnectivityCheckController that discovers the list of endpoints to
+// check by looking for PodNetworkConnectivityChecks in a given namespace, for a specific pod. Updates to
+// the PodNetworkConnectivityCheck status are queued up and handled asynchronously such that disruptions
+// to the ability to update the PodNetworkConnectivityCheck status do not disrupt the ability to perform
+// the connectivity checks.
+type controller struct {
+	factory.Controller
+	podName      string
+	podNamespace string
+	checksGetter operatorcontrolplaneclientv1alpha1.PodNetworkConnectivityCheckInterface
+	checkLister  v1alpha1.PodNetworkConnectivityCheckNamespaceLister
+	recorder     events.Recorder
+	// each PodNetworkConnectivityCheck gets its own PodNetworkConnectivityCheckStatusUpdater
+	updaters map[string]PodNetworkConnectivityCheckStatusUpdater
+}
+
+// Returns a new PodNetworkConnectivityCheckController that performs network connectivity checks
+// as specified in the PodNetworkConnectivityChecks defined in the specified namespace, for the specified pod.
+func New(podName, podNamespace string, checksGetter operatorcontrolplaneclientv1alpha1.PodNetworkConnectivityChecksGetter, checkInformer alpha1.PodNetworkConnectivityCheckInformer, recorder events.Recorder) PodNetworkConnectivityCheckController {
+	c := &controller{
+		podName:      podName,
+		podNamespace: podNamespace,
+		checksGetter: checksGetter.PodNetworkConnectivityChecks(podNamespace),
+		checkLister:  checkInformer.Lister().PodNetworkConnectivityChecks(podNamespace),
+		recorder:     recorder,
+		updaters:     map[string]PodNetworkConnectivityCheckStatusUpdater{},
+	}
+	c.Controller = factory.New().
+		WithSync(c.Sync).
+		WithInformers(checkInformer.Informer()).
```

Comment:
This will cause self-triggering hot looping when you update status, won't it?

Reply:
Done. Used `WithBareInformers` instead.